### PR TITLE
[Merged by Bors] - Retrospective invalidation of exec. payloads for opt. sync

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Adds the functionality to allow blocks to be validated/invalidated after their import as per the [optimistic sync spec](https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md#how-to-optimistically-import-blocks). This means:

- Updating `ProtoArray` to allow flipping the `execution_status` of ancestors/descendants based on payload validity updates.
- Creating separation between `execution_layer` and the `beacon_chain` by creating a `PayloadStatus` struct.
- Refactoring how the `execution_layer` selects a `PayloadStatus` from the multiple statuses returned from multiple EEs.
- Adding testing framework for optimistic imports.
- Add `ExecutionBlockHash(Hash256)` new-type struct to avoid confusion between *beacon block roots* and *execution payload hashes*.
- Add `merge` to [`FORKS`](https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/Makefile#L17) in the `Makefile` to ensure we test the beacon chain with merge settings.
    - Fix some tests here that were failing due to a missing execution layer.

## TODO

- [ ] Balance tests

### CODE REVIEW

**Comment by paulhauner** [comment]
The integration tests are currently failing on https://github.com/ethereum/go-ethereum/issues/24404

**Comment by paulhauner** [comment]
I've just bumped `cc` to 1.0.73 as per https://github.com/alexcrichton/cmake-rs/issues/141.

I think this break is because Github Actions has migrated to Windows 2022 for runners.

**Comment by paulhauner** [comment]
> The integration tests are currently failing on [ethereum/go-ethereum#24404](https://github.com/ethereum/go-ethereum/issues/24404)

That issue has been resolved.

**Comment by paulhauner** [comment]
I just synced this on [`merge-devnet-4`](https://notes.ethereum.org/D5xK4XrmTb6MKGudf_hcrg?view) :tada:

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
            // De-alias 0x00..00 to the genesis block root.
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: maybe this should be a `get_justified_block()` method in fork choice?

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
                "msg" => "ensure you are not connected to a malicious network. This error is not \
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
                    "Unable to trigger client shut down";
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: Do we have to de-alias the genesis block root here too?

File: beacon_node/beacon_chain/src/errors.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/errors.rs:149]: Maybe we want to include the `ExecutionBlockHash` in this error as well as the `JustifiedPayloadInvalid` variant ? Should be available when each are thrown.

File: beacon_node/execution_layer/src/payload_status.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/payload_status.rs:190]: It seems like errors will only surface if there is no valid status? I think it could be useful to log out if there's any errors so it's easier to maintain backup execution engines

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/execution_payload.rs]:
```suggestion
                // Returning an error here should be sufficient to invalidate the block. We have no
```

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/lib.rs]:
```suggestion
                                This has potentially caused a missed block proposal.",
```

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/lib.rs]: Since we are potentially sending this request to multiple execution engines, this `CRIT` might be unnecessarily alarming in cases where fallback works correctly

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
    /// (perhaps it's junk or pre-finalization), then only the `head_block_root` block will be
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
    /// returns `false` if either input root is unknown.
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
    /// - `None` if the latest valid ancestor of `head_block_root` is unknown.
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]: Should this be "unknown"?

```suggestion
                    // ancestor is unknown, exit now with an error.
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]: If I'm understanding this correctly, if the best child is contained in `invalidated_indices`, this field should be set to `None`. So I think this filter condition should be inverted: `!invalidated_indices.contains(best_child)`

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]: This one too

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
                // This block is pre-merge, therefore it has no execution status. Nor do its
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
            // Collect all *descendants* which have been declared invalid since they're the descendant of a block
```

File: consensus/proto_array/src/proto_array.rs
Comment (realbigsean) [inline consensus/proto_array/src/proto_array.rs:138]: Because we apply the proposer boost to this delta later on, it seems possible to turn this value positive again. 

That said, since we added a check for an invalid payload in `node_is_viable_for_head` and since nodes with invalid payloads shouldn't be referenced from other valid nodes I'm not sure we need to handle the delta for these nodes. They should just never be candidates for the head and have no impact on the candidacy of other nodes right?

If we want to set this to zero regardless though, maybe we should just set the `node.weight` to `0` if `node.execution_status.is_invalid()` around line 189.

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs]:
```suggestion
                                "msg" => "No payload ID, the engine is likely syncing. \
                                          This has the potential to cause a missed block proposal.",
```

File: beacon_node/execution_layer/src/test_utils/mod.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/test_utils/mod.rs]: I'm not seeing where this is used, not sure if I'm missing it?

File: beacon_node/beacon_chain/tests/payload_invalidation.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/payload_invalidation.rs]:
```suggestion
/// Ensure that an invalid payload can invalidate its parent too (given the right
```

File: beacon_node/beacon_chain/tests/payload_invalidation.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/payload_invalidation.rs]:
```suggestion
/// Ensure that a `latest_valid_hash` for a pre-finality block only reverts a single block.
```

File: beacon_node/beacon_chain/tests/payload_invalidation.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/payload_invalidation.rs]:
```suggestion
    // Import a pre-finalized block.
```

**Review by realbigsean — CHANGES_REQUESTED** [review]

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
    /// Updates the `verified_node_index` and all ancestors to have validated execution payloads.
```

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]: Like the `Valid` case this should be unreachable too, right? It's impossible for an invalid execution payload to have descendants with execution disabled?

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]: It seems we _will_ return an error in this case: `UnknownLatestValidAncestorHash`?

We could also safely invalidate the descendents of `head_block_root` in this case too, couldn't we? Perhaps before returning the error (or instead?)

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
        // `head_block_root` and the `latest_valid_ancestor_root`.
```

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]: We could start from `head_block_root` here if `latest_valid_ancestor_is_descendant` is `false` (per my previous comment).

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/execution_payload.rs]: Thinking aloud a little, we're relying on an assumption here that `notify_new_payload` is called on blocks _in order_ such that a block _must not_ have any descendants if it is found invalid at this stage.

I think that's safe, but I do wonder whether we could hit something funky in the presence of optimistic sync plus two (slashable) blocks from the same slot with different `BeaconBlockBody`s and identical `ExecutionPayload`s. The first block could be imported optimistically, with other optimistic descendants imported on top of it. Later, we import the slashable block and find that its identical payload is invalid due to the terminal block condition. We _could_ safely invalidate the descendants of the first block at this point, but will not with the code as currently written. I don't think this is a big deal however as the EE will eventually inform us of the invalidity of the descendants if we send it a fork choice update.

Also there's a typo:

```suggestion
                // information to indicate its parent is invalid, so no need to run
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs:3784]: Flagging this as the case that I think should trigger invalidation of the block's descendants per `propagate_execution_payload_invalidation`

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/execution_payload.rs:71]: Passing the parent root here seems strange because the parent is not necessarily invalid. Doing so contradicts this comment on `propagate_execution_payload_invalidation`:

> /// The `head_block_root` is the block root of the latest invalid block.

I think the logic is actually OK because `propagate_execution_payload_invalidation` gracefully handles this case, but maybe we should avoid calling it at all if `block.parent_root() == latest_valid_hash`?

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Nice work! Another important piece for The Merge falling into place :panda_face: 

I tried to think of edge cases and break things, and will be interested to see what you think of the scenarios I mentioned. I may be paranoid/missing some things :sweat_smile: 

> Add merge to [FORKS](https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/Makefile#L17) in the Makefile to ensure we test the beacon chain with merge settings.

I can't see this change, maybe you've got it lurking locally?

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3784]: i.e. this comment https://github.com/sigp/lighthouse/pull/2837#discussion_r813467575

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/execution_payload.rs:71]: Unfortunately we can't do `block.parent_root() == latest_valid_hash` since one is a *block root* and the other a *payload hash*.

File: beacon_node/beacon_chain/src/errors.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/errors.rs:149]: Good idea, added in 1ed7431

File: beacon_node/execution_layer/src/lib.rs
Comment (paulhauner) [reply beacon_node/execution_layer/src/lib.rs]: Good point, lowered to `error` in 5142db0

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Good call, I added it here 1ed7431ba

It turns out we don't need to de-alias the genesis root. Fork choice de-aliases by default, I added some tests to ensure this here ae29684

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: It turns out we don't need to de-alias, as per https://github.com/sigp/lighthouse/pull/2837#discussion_r815243301.

I added a `ForkChoice::get_finalized_block` over in https://github.com/sigp/lighthouse/commit/1ed7431ba2752b3bbefda99811392b138cf0d449.

File: beacon_node/execution_layer/src/payload_status.rs
Comment (paulhauner) [reply beacon_node/execution_layer/src/payload_status.rs:190]: Good call, added in 7eed122.

File: beacon_node/execution_layer/src/test_utils/mod.rs
Comment (paulhauner) [reply beacon_node/execution_layer/src/test_utils/mod.rs]: Nice spotting, removed in 8da2668

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Also fixed in c7d0c6c

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Ooh yeah, good catch! :hot_pepper: I think I was being too fancy here, so I went with something simpler over in c7d0c6c.

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs:138]: Ah yes, great point. I've addressed this over in 4e3a078

Perhaps @michaelsproul could give that commit a once-over too, since he's worked with proposer boost? :pray:

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Yep, you're totally right! In e9905aa I've removed this error and then invalidated the descendants of the head when the latest valid hash isn't available/suitable

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Covered in e9905aa too :+1:

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Ah yes, good point! I added an error in b3be437

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/execution_payload.rs]: > We could safely invalidate the descendants of the first block at this point, but will not with the code as currently written.

AIUI, the solution [in this comment](https://github.com/sigp/lighthouse/pull/2837#discussion_r815261289) has resolved this issue. Is that right?

File: beacon_node/execution_layer/src/lib.rs
Comment (paulhauner) [reply beacon_node/execution_layer/src/lib.rs]: Fixed via https://github.com/sigp/lighthouse/pull/2837#discussion_r813396576

**Comment by paulhauner** [comment]
Thanks @realbigsean and @michaelsproul! That was a great review, thanks for the improvements and bug fixes :grimacing: 

I have addressed all your comments and am waiting for CI to pass :relaxed: 

> I can't see this [`FORKS`] change, maybe you've got it lurking locally?

I must have lost this one between my desktop and laptop, thanks! Added here 2ae6d78.

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]:
```suggestion
                    // descendants of `head_block_root`/`latest_valid_ancestor_root`.
```

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [reply consensus/proto_array/src/proto_array.rs:138]: LGTM!

File: consensus/proto_array/src/proto_array.rs
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]: I think this comment needs to be reworked now that the code has changed.

Specifically

> The `head_block_root` is the block *root* of the latest invalid block.

As discussed [here](https://github.com/sigp/lighthouse/pull/2837#discussion_r813545926) the block root we pass isn't necessarily the block root of an invalid block. It could be the parent of an invalid block.

Here's my attempt at an updated description:

```
Invalidate the relevant ancestors and descendants of a block with an invalid execution payload.

The `head_block_root` should be the beacon block root of the block with the invalid execution payload, _or_ its parent.

The `latest_valid_hash` should be the hash of most recent *valid* execution payload contained in an ancestor block of `head_block_root`.

This function will invalidate:

* The block matching `head_block_root` _unless_ that block has a payload matching `latest_valid_hash`.
* All ancestors of `head_block_root` back to the block with payload matching `latest_valid_hash` (endpoint > exclusive). In the case where the `head_block_root` is the parent of the invalid block and itself matches `latest_valid_hash`, no ancestors will be invalidated.
* All descendants of `latest_valid_hash` if supplied and consistent with `head_block_root`, or else all descendants of `head_block_root`.
```

**Review by michaelsproul — APPROVED** [review]
Looking good! I'm satisfied with the latest updates, and the test coverage

Just a few docs that I think we should update, then we can merge

File: beacon_node/beacon_chain/src/execution_payload.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/execution_payload.rs]: I was imagining that change was already implemented and trying to think of other potential issues. But I think the thing I raised is inconsequential: we might just take one extra call to the EL to work out that we should invalidate a chain that shares an invalid payload with another chain (which is only possible if we have two distinct [slashable] blocks at the same slot). It's a non-issue :relieved:

File: consensus/proto_array/src/proto_array.rs
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: Good point! I've adapted a slightly modified version of your suggestion here: https://github.com/sigp/lighthouse/pull/2837/commits/6dee54c860eefe1bbc5a70d7ebc5353b1dc8f288. Thanks!

**Comment by paulhauner** [comment]
Thanks for the approval @michaelsproul :pray: 

I'll wait for approval/comments from @realbigsean before merging :relaxed:

**Comment by paulhauner** [comment]
I added some non-substantive tests for the `proto_array` crate in https://github.com/sigp/lighthouse/pull/2837/commits/1ab1c433e6968f689ff036965b191adc015b3144.

I was testing to check balance propagation with and without proposer boost. I didn't find any bugs.

**Review by realbigsean — APPROVED** [review]
Ok changes look good to me!

**Comment by paulhauner** [comment]
bors r+

:tada:

**Pull request closed without merging**
